ports/rp2/rp2_dma.c

```diff
@@ -0,0 +1,540 @@
+        return (dma_hw->ints0 & (1u << self->channel)) != 0;
+    }
+    return 0;
+}
+
+STATIC const mp_irq_methods_t rp2_dma_irq_methods = {
+    .trigger = rp2_dma_irq_trigger,
+    .info = rp2_dma_irq_info,
+};
+
+STATIC mp_obj_t rp2_dma_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 0, 0, false);
+
+    int dma_channel = dma_claim_unused_channel(false);
+    if (dma_channel < 0) {
+        mp_raise_OSError(MP_EBUSY);
+    }
+
+    rp2_dma_obj_t *self = m_new_obj(rp2_dma_obj_t);
+    self->base.type = &rp2_dma_type;
+    self->channel = dma_channel;
+
+    // Return the DMA object.
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC void rp2_dma_error_if_closed(rp2_dma_obj_t *self) {
+    if (self->channel == CHANNEL_CLOSED) {
+        mp_raise_ValueError(MP_ERROR_TEXT("Channel closed"));
+    }
+}
+
+STATIC void rp2_dma_attr(mp_obj_t self_in, qstr attr_in, mp_obj_t *dest) {
+    rp2_dma_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (dest[0] == MP_OBJ_NULL) {
+        // Load attribute
+        dma_channel_hw_t *reg_block = dma_channel_hw_addr(self->channel);
+        if (attr_in == MP_QSTR_active) {
+            rp2_dma_error_if_closed(self);
+            uint32_t busy = dma_channel_is_busy(self->channel);
+            dest[0] = mp_obj_new_bool((mp_int_t)busy);
+        } else if (attr_in == MP_QSTR_default_ctrl) {
+            // Get the default ctrl for _this_ channel, i.e. with chaining disabled.
+            rp2_dma_config_obj_t *conf = m_new_obj(rp2_dma_config_obj_t);
+            conf->base.type = &rp2_dma_config_type;
+            conf->value = DEFAULT_DMA_CONFIG | ((self->channel & 0xf) << 11);
+            dest[0] = conf;
+        } else if (attr_in == MP_QSTR_read) {
+            rp2_dma_error_if_closed(self);
+            dest[0] = mp_obj_new_int_from_uint((mp_uint_t)reg_block->read_addr);
+        } else if (attr_in == MP_QSTR_write) {
+            rp2_dma_error_if_closed(self);
+            dest[0] = mp_obj_new_int_from_uint((mp_uint_t)reg_block->write_addr);
+        } else if (attr_in == MP_QSTR_count) {
+            rp2_dma_error_if_closed(self);
+            dest[0] = mp_obj_new_int_from_uint((mp_uint_t)reg_block->transfer_count);
+        } else if (attr_in == MP_QSTR_ctrl) {
+            rp2_dma_error_if_closed(self);
+            dest[0] = mp_obj_new_int_from_uint((mp_uint_t)reg_block->al1_ctrl);
+        } else if (attr_in == MP_QSTR_channel_id) {
+            dest[0] = mp_obj_new_int_from_uint(self->channel);
+        } else if (attr_in == MP_QSTR_registers) {
+            mp_obj_array_t *reg_view = m_new_obj(mp_obj_array_t);
+            mp_obj_memoryview_init(reg_view, 'I', 0, 16, dma_channel_hw_addr(self->channel));
+            dest[0] = reg_view;
+        } else {
+            // If a Micropython class supports attributes then the locals dict is not searched.
+            // We do it manually here.
```

Comment:
In e2ca8ab8fc2f824dc7ebaa8fe65c4d2e7891080e this was improved and you should now be able to return `dest[1] = MP_OBJ_SENTINEL` to continue lookup in `locals_dict`.

Reply:
Aha! That's very helpful new functionality since I wrote this code. I will excise a the locals looks code, which might go some way to your request to _"retain the same functionality but in half the code size"_.